File: languages/nodejs/src/benchmarkers/benchmarker-manager.ts
import { Benchmarker } from '@benchmarkers/benchmarker';
import { QuickSortBenchmarker } from '@benchmarkers/impl/quick-sort-benchmarker';

type BenchmarkerMap = {[id: string]: Benchmarker};

class BenchmarkerManager {
    benchmarkers: BenchmarkerMap;

    constructor() {
        const benchmarkerList = [
            new QuickSortBenchmarker(),
        ];

        this.benchmarkers = benchmarkerList
            .reduce((map, benchmarker) => {
                map[benchmarker.operationName] = benchmarker;
                return map;
            }, {} as BenchmarkerMap);
    }

    getOperationBenchmarker(operationName: string) {
        if (this.benchmarkers[operationName]) {
            return this.benchmarkers[operationName];
        } else {
            const validNames = this.getValidOperations().join(', ');
            throw new Error(`Operation name '${operationName}' not found. Operation must be one of the following: ${validNames}`);
        }
    }

    private getValidOperations() {
        return Object.keys(this.benchmarkers).sort();
    }
}

export const BENCHMARKER_MANAGER = new BenchmarkerManager();

File: languages/c#/src/benchmarkers/BenchmarkerManager.cs
namespace Benchmarkers
{
    public sealed class BenchmarkerManager
    {
        private static readonly Lazy<BenchmarkerManager> lazy = new(() => new BenchmarkerManager());

        public static BenchmarkerManager Instance { get { return lazy.Value; } }

        private readonly Dictionary<string, Benchmarker> benchmarkers;

        private BenchmarkerManager()
        {
            List<Benchmarker> benchmarkerList = [
                new QuickSortBenchmarker()
            ];
            benchmarkers = benchmarkerList.ToDictionary(b => b.OperationName, b => b);
        }

        public Benchmarker GetOperationBenchmarker(string operationName) {
            if (benchmarkers.TryGetValue(operationName, out Benchmarker? value)) {
                return value;
            } else {
                string validNames = string.Join(", ", GetValidOperations());
                throw new InvalidOperationException(
                    $"Operation name '{operationName}' not found. Operation must be one of the following: {validNames}"
                );
            }
        }

        private List<string> GetValidOperations()
        {
            List<string> validOperations = [.. benchmarkers.Keys];
            validOperations.Sort();
            return validOperations;
        }
    }
}